### PR TITLE
Add SciKitLearn regression metric into RAVEN metric system

### LINKED ISSUE #396: Add SciKitLearn regression metrics inside RAVEN Metric system, and some modifications in Cross Validation
--------
Issue Description
--------
##### What did you expect to see happen?
Add some new metrics, and some modifications in RAVEN Metrics and Cross Validation PP
Raise an error when the input keywords is not expected by the Metrics. 
Only allow to use regression metrics inside Cross Validation PP, and rename the outputs of metric results to enable the use of the same type of metrics with different names in the Metric PP and Cross Validation PP.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [ ] 1. Is it tagged with a type: defect or improvement?
- [ ] 2. Is it tagged with a priority: critical, normal or minor?
- [ ] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [ ] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [ ] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
close #396 
Add regression metrics inside the Metric system, and only allow the Cross Validation PP to use regression metrics
##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/CodeInterfaces/RAVEN/RAVENInterface.py [hunk 1]
```diff
@@ -264,19 +264,20 @@ def checkForOutputFailure(self,output,workingDir):
...
-       try:
-         fileObj = open(outstreamFile,"r")
-       except IOError:
-         print(self.printTag+' ERROR: The RAVEN SLAVE output file "'+str(outstreamFile)+'" does not exist!')
-         failure = True
-       if not failure:
-         readLines = fileObj.readlines()
-         if any("nan" in x.lower() for x in readLines):
+      for filename in self.linkedDataObjectOutStreamsNames:
+        outstreamFile = os.path.join(workingDir,self.innerWorkingDir,filename+".csv")
```
Comment (wangcj05) [inline framework/CodeInterfaces/RAVEN/RAVENInterface.py]: should be ``outStreamFile``?

File: framework/JobHandler.py [hunk 2]
```diff
@@ -143,11 +143,12 @@ def __checkAndRemoveFinished(self, running):
       returnCode = running.getReturnCode()
       if returnCode != 0:
         metadataFailedRun = running.getMetadata()
-        metaDataKeys      = metadataFailedRun.keys()
         metaDataToKeep    = metadataFailedRun
-        if 'jobHandler' in metaDataKeys:
-          metaDataKeys.pop(metaDataKeys.index("jobHandler"))
-          metaDataToKeep = { keepKey: metadataFailedRun[keepKey] for keepKey in metaDataKeys }
+        if metadataFailedRun is not None:
```
Comment (wangcj05) [inline framework/JobHandler.py]: for consistent, either use ``metadata*`` or ``metaData*``.

File: framework/PostProcessors/CrossValidation.py [hunk 3]
```diff
@@ -360,3 +373,49 @@ def _writeText(self,output,outputDictionary, separator=' '):
             output.write(separator+str(valueDict[metricName][cvRunNum]))
         output.write(os.linesep)
 
+  def _writeDataObject(self,output,outputDictionary):
+    """
+      Defines the method for writing the post-processor to a a PointSet
+      @ In, output, PointSet object, PointSet to dumpt the results to
```
Comment (wangcj05) [inline framework/PostProcessors/CrossValidation.py]: change ``dumpt`` to ``dump``

**Comment by wangcj05** [comment]
I have finished the review. Some minor changes, do not forget to add the tests to test all the new metrics and test the ``_writeDataObject`` method, as well as documenting the changes in the user manual.

**Comment by wangcj05** [comment]
Please let me know when this PR is ready for a second round of review.

**Comment by wangcj05** [comment]
@alfoa FYI, I will start to work on this PR.

**Comment by wangcj05** [comment]
@alfoa I think this PR is ready for reviewing. Regarding your questions: 

- Why do I need to specify the number of samples "n" in CrossValidation PostProcessor?
Answer: This is required for the SciKit-Learn version 0.17-0.19 if we import the cross_validation module . In the future, if we switch to import cross_validation from the model_selection, we do not need the number of samples "n".
- If I input the same metric type (with different attribute name) I get a crash in the CrossValidation PP but not a caught error
Answer: This is because there is no method available in the SciKit-Learn cross_validation module to obtain the names of input parameters.

**Comment by wangcj05** [comment]
@alfoa Please let me know what else you need me to add.

**Comment by wangcj05** [comment]
@PaulTalbot-INL Do you have time to take a look at this PR?

File: framework/PostProcessors/CrossValidation.py [hunk 4]
```diff
@@ -107,10 +108,18 @@ def _localReadMoreXML(self, xmlNode):
...
     self.initializationOptionDict = {}
-
+    average = None
+    cvNode = xmlNode.find('SciKitLearn')
+    for child in cvNode:
+      if child.tag == 'average':
+        average = child
+        break
+    if average is not None:
+      self.averageScores = average.text.lower() in utils.stringsThatMeanTrue()
```
Comment (PaulTalbot-INL) [inline framework/PostProcessors/CrossValidation.py]: Correct me if I'm wrong, but couldn't this be simplified to:
```python
if child.tag == 'average':
  self.averageScores = child.text.lower() in utils.stringsThatMeanTrue()
```
and eliminate the need for the `if average is not None` line?

**Comment by wangcj05** [comment]
Regarding Andrea's concern, currently I do not have a simple solution for that.

**Comment by wangcj05** [comment]
I think let's open an issue for that, and fix that in a different PR. @alfoa @PaulTalbot-INL

**Comment by PaulTalbot-INL** [comment]
Maybe I don't understand, but isn't it a matter of catching the error thrown when you change metrics but leave the attributes the same?

**Comment by PaulTalbot-INL** [comment]
Can we also edit the header of this PR to explain what this set of changes accomplishes?  Thanks.  We'll be grateful later.

**Comment by alfoa** [comment]
@wangcj05:   
Why do I need to specify the number of samples "n" in CrossValidation PostProcessor?
Answer: This is required for the SciKit-Learn version 0.17-0.19 if we import the cross_validation module . In the future, if we switch to import cross_validation from the model_selection, we do not need the number of samples "n".
_**I agree it is required by sklearn, but the number of samples can be inferred from the DataObjects directly in the PP (len(DataObject)). Am I missing something here?**_

If I input the same metric type (with different attribute name) I get a crash in the CrossValidation PP but not a caught error
Answer: This is because there is no method available in the SciKit-Learn cross_validation module to obtain the names of input parameters.
_** I think I did not explain myself. If I input 
```
  <Metrics>
    <SKL name="wolf1">
      <metricType>euclidean</metricType>
    </SKL>
    <SKL name="wolf2">
      <metricType>euclidean</metricType>
    </SKL>
  </Metrics>
```
And use them in either the Metric PP or CrossValidation the code crashes**_

**Comment by wangcj05** [comment]
@alfoa Regarding the parameter 'n', I agree with you, we can remove it. I have the same issue in the Hybrid Model, what I did in the Hybrid Model is to reset the parameter 'n' for the updated data objects. 

Regarding the crashes, sorry for the misunderstanding, and I will fix it.

**Comment by PaulTalbot-INL** [comment]
Whoops, I only wanted to approve that one commit.  Apparently that approve all of it.

**Comment by wangcj05** [comment]
@PaulTalbot-INL Now, I think I have resolved all the issues, and the tests have been modified to test the changes.
